core/modules/replica/IndexJob.cc

```diff
@@ -0,0 +1,538 @@
+        controller()->serviceProvider()->config()->workers();
+
+    // Initialize a collection of chunks grouped by workers, in a way which
+    // would make a best attempt to keep requests equally (as much as that's possible)
+    // balanced between the workers.
+    //
+    // Note, that the algorithm considers a possibility that chunks may have
+    // multiple (more than 1) replicas. In this scenario a choice which replica
+    // of a particular chunk to use will be based on the number of the chunk's
+    // replicas as well as on the total number of chunks per each worker.
+
+    // The first step is to find workers which store replicas of each
+    // chunk to be processed by the job.
+
+    bool const unusedAllDatabases = false;  // is still required by the method's contract
+    bool const unusedIsPublished = true;    // is still required by the method's contract
+    bool const includeFileInfo = false;     // to speed up the query as we don't need file info
+
+    map<unsigned int, list<string>> chunk2workers;
+    for (auto&& worker: workerNames) {
+        vector<ReplicaInfo> replicas;
+        controller()->serviceProvider()->databaseServices()->findWorkerReplicas(
+            replicas,
+            worker,
+            database(),
+            unusedAllDatabases,
+            unusedIsPublished,
+            includeFileInfo
+        );
+        for (auto&& replica: replicas) {
+            chunk2workers[replica.chunk()].push_back(replica.worker());
+        }
+    }
+    
+    // Now build the plan for each worker based on the above harvested
+    // distribution of chunk replicas across workers.
+    //
+    // TODO: this single-pass algorithm may be biased to an order
+    // in which chunks are being processed by the algorithm. Consider
+    // a more sophisticated implementation which would be bias-free.
+
+    for (auto&& itr: chunk2workers) {
+
+        unsigned int const chunk = itr.first;
+        auto&& workers = itr.second;
+
+        // Find the least loaded worker from those where chunk replicas
+        // are residing.
+        string worker;
+        size_t minNumChunks = numeric_limits<size_t>::max();
+        for (auto&& candidateWorker: workers) {
+            auto const numChunksAtCandidate = _chunks[candidateWorker].size();
+            if (numChunksAtCandidate < minNumChunks) {
+                worker = candidateWorker;
+                minNumChunks = numChunksAtCandidate;
+            }
+        }
+        if (worker.empty()) {
+            throw logic_error(context() + string(__func__) + ":  internal bug");
+        }
+        _chunks[worker].push_back(chunk);
+    }
+
+    // Launch the initial batch of requests in the number which won't exceed
+    // the number of the service processing threads at each worker multiplied
+    // by the number of workers involved into the operation and by the "magic"
+    // number 8. The later is needed to absorb the latency of the network
```

Comment:
I accept the critics. However this code is still in the "under development" case. It will improve later. I was playing with different scaling factors to get an idea how much this would affect the overall performance of the "secondary index" building algorithm. The code will be finalized in some later iteration of the `Ingest` system.